sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/exchange/UsdExchangeRateFetcher.kt

```diff
@@ -0,0 +1,68 @@
+package cash.z.ecc.android.sdk.internal.exchange
+
+import cash.z.ecc.android.sdk.internal.Twig
+import cash.z.ecc.android.sdk.internal.model.TorClient
+import cash.z.ecc.android.sdk.model.FetchFiatCurrencyResult
+import cash.z.ecc.android.sdk.model.FiatCurrencyConversion
+import kotlinx.coroutines.delay
+import kotlinx.datetime.Clock
+import java.io.File
+
+internal class UsdExchangeRateFetcher(private val torDir: File) {
+    @Suppress("TooGenericExceptionCaught", "ReturnCount")
+    suspend operator fun invoke(): FetchFiatCurrencyResult =
+        retry {
+            val tor =
+                try {
+                    Twig.info { "[USD] Tor client bootstrap" }
+                    TorClient.new(torDir)
+                } catch (e: Exception) {
+                    Twig.error(e) { "[USD] To client bootstrap failed" }
+                    return@retry FetchFiatCurrencyResult.Error(exception = e)
+                }
+
+            val rate =
+                try {
+                    Twig.info { "[USD] Fetch start" }
+                    tor.getExchangeRateUsd()
+                } catch (e: Exception) {
+                    Twig.error(e) { "[USD] Fetch failed" }
+                    return@retry FetchFiatCurrencyResult.Error(e)
+                } finally {
+                    tor.dispose()
+                }
+
+            Twig.debug { "[USD] Fetch success: $rate" }
```

Comment:
If an error occurs in `tor.getExchangeRateUsd()`, the retry logic will discard the `tor` client and create a new one with `TorClient.new`. This is going to noticeably slow down the retry logic, as on each retry it is forcing the Tor client to:
- serialize its state
- shut down its circuits
- load it state
- build new circuits, which can take several seconds

This should instead have separate retry logic for `TorClient.new` (trying a few times to create the Tor client) and `tor.getExchangeRateUsd()` (trying a few times in case a particular Tor request failed, but reusing the already-running Tor client).

Reply:
good catch